File: src/tp04/ej03/Rectangle.java
package tp04.ej03;

public class Rectangle extends Figure {
    private Point corner1, corner2;

    public Rectangle(Point corner1, Point corner2) {
        this.corner1 = corner1;
        this.corner2 = corner2;
    }

    public double perimeter() {
        return 2 * base() + 2 * height();
    }

    public double area() {
        return base() * height();
    }

    public double base() {
        return Math.abs(corner1.getX() - corner2.getX());
    }

    public double height() {
        return Math.abs(corner1.getY() - corner2.getY());
    }

    @Override
    public String toString() {
       return "Rectangulo [ " + corner1 + " , " + corner2 + " ]";
    }

    @Override
    public void moveNorth(double delta) {
        corner1.moveNorth(delta);
        corner2.moveNorth(delta);
    }

    @Override
    public void moveSouth(double delta) {
        corner1.moveSouth(delta);
        corner2.moveSouth(delta);
    }

    @Override
    public void moveWest(double delta) {
        corner1.moveWest(delta);
        corner2.moveWest(delta);
    }

    @Override
    public void moveEast(double delta) {
        corner1.moveEast(delta);
        corner2.moveEast(delta);
    }
}

File: src/tp08/ej09/rectangle.rb
class Rectangle < Figure
  attr_reader :corner1, :corner2

  def initialize(corner1, corner2)
    @corner1 = corner1
    @corner2 = corner2
  end

  def perimeter
    2 * (@corner1.x - @corner2.x).abs + 2 * (@corner1.y - @corner2.y).abs
  end

  def area
    (@corner1.x - @corner2.x).abs * (@corner1.y - @corner2.y).abs
  end 

  def ==(other)
    return false unless other.is_a? Rectangle
    @corner1 == other.corner1 && @corner2 == other.corner2
  end

  def to_s
    "Rectangle [ #{@corner1} #{@corner2} ]"
  end
end
